hathor/builder.py

```diff
@@ -36,11 +40,266 @@
+        self._wallet_unlock: Optional[bytes] = None
+
+        self._enable_address_index: bool = False
+        self._enable_tokens_index: bool = False
+        self._enable_utxo_index: bool = False
+
+        self._enable_sync_v1: Optional[bool] = None
+        self._enable_sync_v2: Optional[bool] = None
+
+        self._stratum_port: Optional[int] = None
+
+    def build(self) -> BuildArtifacts:
+        if self.artifacts is not None:
+            return self.artifacts
+
+        settings = self._get_settings()
+        reactor = self._create_reactor()
+        pubsub = self._create_pubsub(reactor)
+
+        peer_id = self._get_peer_id()
+
+        soft_voided_tx_ids = set(settings.SOFT_VOIDED_TX_IDS)
+        consensus_algorithm = ConsensusAlgorithm(soft_voided_tx_ids, pubsub)
+
+        wallet = self._create_wallet()
+        tx_storage = self._create_tx_storage()
+        indexes = tx_storage.indexes
+        assert indexes is not None
+
+        if self._enable_address_index:
+            indexes.enable_address_index(pubsub)
+
+        if self._enable_tokens_index:
+            indexes.enable_tokens_index()
+
+        if self._enable_utxo_index:
+            indexes.enable_utxo_index()
+
+        kwargs: Dict[str, Any] = {}
+
+        if self._enable_sync_v1 is not None:
+            kwargs['enable_sync_v1'] = self._enable_sync_v1
+
+        if self._enable_sync_v2 is not None:
+            kwargs['enable_sync_v2'] = self._enable_sync_v2
+
+        if self._stratum_port is not None:
+            kwargs['stratum_port'] = self._stratum_port
+
+        if self._network is None:
+            raise TypeError('you must set a network')
+
+        if self._event_manager is not None:
+            kwargs['event_manager'] = self._event_manager
+
+        manager = HathorManager(
+            reactor,
+            pubsub=pubsub,
+            consensus_algorithm=consensus_algorithm,
+            peer_id=peer_id,
+            tx_storage=tx_storage,
+            network=self._network,
+            wallet=wallet,
+            rng=self._rng,
+            checkpoints=self._checkpoints,
+            capabilities=self._capabilities,
+            environment_info=get_environment_info(self._cmdline, peer_id.id),
+            **kwargs
+        )
+
+        self.artifacts = BuildArtifacts(
+            settings=settings,
+            rng=self._rng,
+            reactor=reactor,
+            manager=manager,
+            pubsub=pubsub,
+            consensus=consensus_algorithm,
+            tx_storage=tx_storage,
+            indexes=indexes,
+            wallet=wallet,
+            rocksdb_storage=self._rocksdb_storage,
+        )
+
+        return self.artifacts
+
+    def set_event_manager(self, event_manager: EventManager) -> None:
+        self._event_manager = event_manager
+
+    def _get_peer_id(self) -> PeerId:
+        if self._peer_id is not None:
+            return self._peer_id
+        raise NotImplementedError
```

Comment:
I think `raise ValueError('property not set')` would be better than `raise NotImplementedError`, for this and all `_get_*` methods that do this.

Reply:
Done!